src/document.c

```diff
@@ -0,0 +1,367 @@
+void Document_Init(Document *doc, RedisModuleString *docKey, double score, int numFields,
+                   const char *lang, const char *payload, size_t payloadSize) {
+  doc->docKey = docKey;
+  doc->score = (float)score;
+  doc->numFields = numFields;
+  doc->fields = calloc(doc->numFields, sizeof(DocumentField));
+  doc->language = lang;
+  doc->payload = payload;
+  doc->payloadSize = payloadSize;
+}
+
+void Document_Detatch(Document *doc, RedisModuleCtx *srcCtx) {
+  RedisModule_RetainString(srcCtx, doc->docKey);
+  for (size_t ii = 0; ii < doc->numFields; ++ii) {
+    DocumentField *f = doc->fields + ii;
+    RedisModule_RetainString(srcCtx, f->text);
+    f->name = strdup(f->name);
+  }
+  doc->stringOwner = 1;
+  if (doc->payload) {
+    doc->payload = strndup(doc->payload, doc->payloadSize);
+  }
+  if (doc->language) {
+    doc->language = strdup(doc->language);
+  }
+}
+
+void Document_Free(Document *doc) {
+  free(doc->fields);
+}
+
+void Document_FreeDetatched(Document *doc, RedisModuleCtx *anyCtx) {
+  RedisModule_FreeString(anyCtx, doc->docKey);
+
+  for (size_t ii = 0; ii < doc->numFields; ++ii) {
+    RedisModule_FreeString(anyCtx, doc->fields[ii].text);
+    free((void *)doc->fields[ii].name);
+  }
+
+  free((char *)doc->payload);
+  free((char *)doc->language);
+
+  Document_Free(doc);
+}
+
+int Redis_SaveDocument(RedisSearchCtx *ctx, Document *doc) {
+
+  RedisModuleKey *k =
+      RedisModule_OpenKey(ctx->redisCtx, doc->docKey, REDISMODULE_WRITE | REDISMODULE_READ);
+  if (k == NULL || (RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_EMPTY &&
+                    RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_HASH)) {
+    return REDISMODULE_ERR;
+  }
+
+  for (int i = 0; i < doc->numFields; i++) {
+    RedisModule_HashSet(k, REDISMODULE_HASH_CFIELDS, doc->fields[i].name, doc->fields[i].text,
+                        NULL);
+  }
+  return REDISMODULE_OK;
+}
+
+static int Document_Store(Document *doc, RedisSearchCtx *ctx, int options,
+                          const char **errorString) {
+  const char *keystr = RedisModule_StringPtrLen(doc->docKey, NULL);
+  DocTable *table = &ctx->spec->docs;
+
+  // if we're in replace mode, first we need to try and delete the older version of the document
+  if (options & DOCUMENT_ADD_REPLACE) {
+    DocTable_Delete(table, keystr);
+  }
+  doc->docId = DocTable_Put(table, keystr, doc->score, 0, doc->payload, doc->payloadSize);
+
+  // Make sure the document is not already in the index - it needs to be
+  // incremental!
+  if (doc->docId == 0) {
+    *errorString = "Couldn't allocate new document table entry";
+    return -1;
+  }
+
+  if (!(options & DOCUMENT_ADD_NOSAVE)) {
+    // first save the document as hash
+    if (REDISMODULE_ERR == Redis_SaveDocument(ctx, doc)) {
+      *errorString = "Couldn't save document";
+      goto fail;
+    }
+  }
+
+  return 0;
+
+fail:
+  DocTable_Delete(table, keystr);
+  return -1;
+}
+
+typedef struct {
+  ForwardIndex *idx;
+  RSSortingVector *sv;
```

Comment:
The sorting vector does not become shared until later on, (i.e. until `DocTable_SetSortingVector` is called), and once it's called, it is no longer touched - therefore there is no need in protecting it.